File: spec-sdk-tests/tests/destinations/webhook-field-removal.test.ts
import { describe, it, before, after } from 'mocha';
import { expect } from 'chai';
import { SdkClient, createSdkClient } from '../../utils/sdk-client';
import { createWebhookDestination } from '../../factories/destination.factory';
/* eslint-disable no-console */
/* eslint-disable no-undef */

describe('Webhook Destinations - Filter/Metadata/DeliveryMetadata Removal (SDK)', () => {
  let client: SdkClient;
  let destinationId: string;

  before(async () => {
    client = createSdkClient();
    try {
      await client.upsertTenant();
    } catch (error) {
      console.warn('Failed to create tenant (may already exist):', error);
    }
  });

  after(async () => {
    try {
      if (destinationId) {
        await client.deleteDestination(destinationId);
      }
    } catch (error) {
      console.warn('Failed to cleanup:', error);
    }
    try {
      await client.deleteTenant();
    } catch (error) {
      console.warn('Failed to delete tenant:', error);
    }
  });

  describe('Filter removal', () => {
    before(async () => {
      const dest = await client.createDestination(
        createWebhookDestination({
          filter: {
            body: { user_id: 'usr_123' },
          },
        })
      );
      destinationId = dest.id;
    });

    it('should have filter set after creation', async () => {
      const dest = await client.getDestination(destinationId);
      expect(dest.filter).to.not.be.null;
      expect(dest.filter).to.deep.include({ body: { user_id: 'usr_123' } });
    });

    it('should clear filter when set to empty object {}', async () => {
      const updated = await client.updateDestination(destinationId, {
        filter: {},
      });
      const isEmpty =
        updated.filter === null ||
        updated.filter === undefined ||
        (typeof updated.filter === 'object' && Object.keys(updated.filter).length === 0);
      expect(isEmpty).to.be.true;
    });

    it('should not change filter when field is omitted', async () => {
      await client.updateDestination(destinationId, {
        filter: { body: { user_id: 'usr_456' } },
      });
      const updated = await client.updateDestination(destinationId, {
        topics: ['user.created', 'user.updated'],
      });
      expect(updated.filter).to.deep.include({ body: { user_id: 'usr_456' } });
    });
  });

  describe('Metadata removal', () => {
    before(async () => {
      try {
        if (destinationId) await client.deleteDestination(destinationId);
      } catch {}
      const dest = await client.createDestination(
        createWebhookDestination({
          metadata: {
            env: 'production',
            team: 'platform',
            region: 'us-east-1',
          },
        })
      );
      destinationId = dest.id;
    });

    it('should have metadata set after creation', async () => {
      const dest = await client.getDestination(destinationId);
      expect(dest.metadata).to.deep.equal({
        env: 'production',
        team: 'platform',
        region: 'us-east-1',
      });
    });

    it('should remove a single metadata key when sending subset', async () => {
      const updated = await client.updateDestination(destinationId, {
        metadata: {
          env: 'production',
          team: 'platform',
        },
      });
      expect(updated.metadata).to.deep.equal({
        env: 'production',
        team: 'platform',
      });
      expect(updated.metadata).to.not.have.property('region');
    });

    it('should clear all metadata when set to empty object {}', async () => {
      const updated = await client.updateDestination(destinationId, {
        metadata: {},
      });
      const isEmpty =
        updated.metadata === null ||
        updated.metadata === undefined ||
        (typeof updated.metadata === 'object' && Object.keys(updated.metadata).length === 0);
      expect(isEmpty).to.be.true;
    });

    it('should not change metadata when field is omitted', async () => {
      await client.updateDestination(destinationId, {
        metadata: { env: 'staging' },
      });
      const updated = await client.updateDestination(destinationId, {
        topics: ['user.created', 'user.updated'],
      });
      expect(updated.metadata).to.deep.equal({ env: 'staging' });
    });
  });

  describe('DeliveryMetadata removal', () => {
    before(async () => {
      try {
        if (destinationId) await client.deleteDestination(destinationId);
      } catch {}
      const dest = await client.createDestination(
        createWebhookDestination({
          deliveryMetadata: {
            source: 'outpost',
            version: '1.0',
          },
        })
      );
      destinationId = dest.id;
    });

    it('should have delivery_metadata set after creation', async () => {
      const dest = await client.getDestination(destinationId);
      expect(dest.deliveryMetadata).to.deep.equal({
        source: 'outpost',
        version: '1.0',
      });
    });

    it('should remove a single delivery_metadata key when sending subset', async () => {
      const updated = await client.updateDestination(destinationId, {
        deliveryMetadata: {
          source: 'outpost',
        },
      });
      expect(updated.deliveryMetadata).to.deep.equal({
        source: 'outpost',
      });
      expect(updated.deliveryMetadata).to.not.have.property('version');
    });

    it('should clear all delivery_metadata when set to empty object {}', async () => {
      const updated = await client.updateDestination(destinationId, {
        deliveryMetadata: {},
      });
      const isEmpty =
        updated.deliveryMetadata === null ||
        updated.deliveryMetadata === undefined ||
        (typeof updated.deliveryMetadata === 'object' &&
          Object.keys(updated.deliveryMetadata).length === 0);
      expect(isEmpty).to.be.true;
    });

    it('should not change delivery_metadata when field is omitted', async () => {
      await client.updateDestination(destinationId, {
        deliveryMetadata: { source: 'test' },
      });
      const updated = await client.updateDestination(destinationId, {
        topics: ['user.created', 'user.updated'],
      });
      expect(updated.deliveryMetadata).to.deep.equal({ source: 'test' });
    });
  });
});

File: spec-sdk-tests/tests/destinations/webhook-field-removal_test.go
package destinations_test

import (
	"context"
	"os"
	"testing"

	outpostgo "github.com/hookdeck/outpost/sdks/outpost-go"
	"github.com/hookdeck/outpost/sdks/outpost-go/models/components"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func getClient(t *testing.T) *outpostgo.Outpost {
	t.Helper()
	apiKey := os.Getenv("API_KEY")
	if apiKey == "" {
		apiKey = "apikey"
	}
	baseURL := os.Getenv("API_BASE_URL")
	if baseURL == "" {
		baseURL = "http://localhost:3333/api/v1"
	}
	return outpostgo.New(
		outpostgo.WithSecurity(apiKey),
		outpostgo.WithServerURL(baseURL),
	)
}

const testTenantID = "go-sdk-test-tenant"

func setupTenant(t *testing.T, client *outpostgo.Outpost) {
	t.Helper()
	ctx := context.Background()
	_, err := client.Tenants.Upsert(ctx, testTenantID, nil)
	require.NoError(t, err, "failed to upsert tenant")
}

func cleanupTenant(t *testing.T, client *outpostgo.Outpost) {
	t.Helper()
	ctx := context.Background()
	_, _ = client.Tenants.Delete(ctx, testTenantID)
}

func createWebhookDest(t *testing.T, client *outpostgo.Outpost, filter map[string]any, metadata, deliveryMetadata map[string]string) string {
	t.Helper()
	ctx := context.Background()
	topics := components.CreateTopicsArrayOfStr([]string{"user.created", "user.updated"})
	resp, err := client.Destinations.Create(ctx, testTenantID, components.CreateDestinationCreateWebhook(components.DestinationCreateWebhook{
		Topics: topics,
		Config: components.WebhookConfig{
			URL: "https://example.com/webhook",
		},
		Filter:           filter,
		Metadata:         metadata,
		DeliveryMetadata: deliveryMetadata,
	}))
	require.NoError(t, err, "failed to create destination")
	wh := resp.GetDestinationWebhook()
	require.NotNil(t, wh, "expected webhook destination in response")
	return wh.ID
}

func deleteDest(t *testing.T, client *outpostgo.Outpost, destID string) {
	t.Helper()
	ctx := context.Background()
	_, _ = client.Destinations.Delete(ctx, testTenantID, destID)
}

func getDest(t *testing.T, client *outpostgo.Outpost, destID string) *components.DestinationWebhook {
	t.Helper()
	ctx := context.Background()
	resp, err := client.Destinations.Get(ctx, testTenantID, destID)
	require.NoError(t, err, "failed to get destination")
	wh := resp.Destination.DestinationWebhook
	require.NotNil(t, wh, "expected webhook destination")
	return wh
}

func updateDest(t *testing.T, client *outpostgo.Outpost, destID string, update components.DestinationUpdateWebhook) *components.DestinationWebhook {
	t.Helper()
	ctx := context.Background()
	resp, err := client.Destinations.Update(ctx, testTenantID, destID,
		components.CreateDestinationUpdateDestinationUpdateWebhook(update))
	require.NoError(t, err, "failed to update destination")
	wh := resp.OneOf.Destination.DestinationWebhook
	require.NotNil(t, wh, "expected webhook destination in update response")
	return wh
}

func TestFilterRemoval(t *testing.T) {
	client := getClient(t)
	setupTenant(t, client)
	defer cleanupTenant(t, client)

	filter := map[string]any{"body": map[string]any{"user_id": "usr_123"}}
	destID := createWebhookDest(t, client, filter, nil, nil)
	defer deleteDest(t, client, destID)

	t.Run("should have filter set after creation", func(t *testing.T) {
		dest := getDest(t, client, destID)
		assert.NotNil(t, dest.Filter)
	})

	t.Run("should clear filter when set to empty map", func(t *testing.T) {
		dest := updateDest(t, client, destID, components.DestinationUpdateWebhook{
			Filter: map[string]any{},
		})
		assert.True(t, dest.Filter == nil || len(dest.Filter) == 0,
			"filter should be nil or empty, got: %v", dest.Filter)
	})
}

func TestMetadataRemoval(t *testing.T) {
	client := getClient(t)
	setupTenant(t, client)
	defer cleanupTenant(t, client)

	metadata := map[string]string{
		"env":    "production",
		"team":   "platform",
		"region": "us-east-1",
	}
	destID := createWebhookDest(t, client, nil, metadata, nil)
	defer deleteDest(t, client, destID)

	t.Run("should have metadata set after creation", func(t *testing.T) {
		dest := getDest(t, client, destID)
		assert.Equal(t, metadata, dest.Metadata)
	})

	t.Run("should remove a single metadata key when sending subset", func(t *testing.T) {
		dest := updateDest(t, client, destID, components.DestinationUpdateWebhook{
			Metadata: map[string]string{
				"env":  "production",
				"team": "platform",
			},
		})
		assert.Equal(t, map[string]string{
			"env":  "production",
			"team": "platform",
		}, dest.Metadata)
		_, hasRegion := dest.Metadata["region"]
		assert.False(t, hasRegion, "region key should have been removed")
	})

	t.Run("should clear all metadata when set to empty map", func(t *testing.T) {
		dest := updateDest(t, client, destID, components.DestinationUpdateWebhook{
			Metadata: map[string]string{},
		})
		assert.True(t, dest.Metadata == nil || len(dest.Metadata) == 0,
			"metadata should be nil or empty, got: %v", dest.Metadata)
	})
}

func TestDeliveryMetadataRemoval(t *testing.T) {
	client := getClient(t)
	setupTenant(t, client)
	defer cleanupTenant(t, client)

	dm := map[string]string{
		"source":  "outpost",
		"version": "1.0",
	}
	destID := createWebhookDest(t, client, nil, nil, dm)
	defer deleteDest(t, client, destID)

	t.Run("should have delivery_metadata set after creation", func(t *testing.T) {
		dest := getDest(t, client, destID)
		assert.Equal(t, dm, dest.DeliveryMetadata)
	})

	t.Run("should remove a single delivery_metadata key when sending subset", func(t *testing.T) {
		dest := updateDest(t, client, destID, components.DestinationUpdateWebhook{
			DeliveryMetadata: map[string]string{
				"source": "outpost",
			},
		})
		assert.Equal(t, map[string]string{
			"source": "outpost",
		}, dest.DeliveryMetadata)
		_, hasVersion := dest.DeliveryMetadata["version"]
		assert.False(t, hasVersion, "version key should have been removed")
	})

	t.Run("should clear all delivery_metadata when set to empty map", func(t *testing.T) {
		dest := updateDest(t, client, destID, components.DestinationUpdateWebhook{
			DeliveryMetadata: map[string]string{},
		})
		assert.True(t, dest.DeliveryMetadata == nil || len(dest.DeliveryMetadata) == 0,
			"delivery_metadata should be nil or empty, got: %v", dest.DeliveryMetadata)
	})
}
